File: src/host/acp/sessionFailure.ts
import type { FailureAction, FailureCategory, TurnError } from '@shared/transcript';

// JetBrains AIR sessionFailure extension (codex-acp / claude-agent-acp): structured warnings and
// errors the adapter reports at `_meta.jetbrains.air.sessionFailure`, on session_info_update
// notifications and on the session/prompt response. Same id + higher revision updates in place;
// a same or lower revision is a duplicate the sender already knows about. Actions are what the
// adapter offers, never inferred from the category.
export interface SessionFailure {
  id: string;
  revision: number;
  category: FailureCategory;
  severity: 'warning' | 'error';
  title: string;
  details?: string;
  actions: FailureAction[];
}

const CATEGORIES = new Set<FailureCategory>(['connection', 'access', 'limit', 'request', 'service', 'unknown']);
const ACTIONS = new Set<FailureAction>(['retry', 'login', 'new_session']);

const text = (v: unknown): string | undefined => typeof v === 'string' && v.trim() ? v.trim() : undefined;

// The strict read of one payload: required fields missing or malformed → undefined (the caller logs).
// An unrecognized category keeps the rest of the payload and degrades to 'unknown'; unknown actions
// are filtered, never guessed
export function failureOf(meta: unknown, log?: (line: string) => void): SessionFailure | undefined {
  const air = (meta as { jetbrains?: { air?: Record<string, unknown> } } | null | undefined)?.jetbrains?.air;
  const raw = air?.sessionFailure;
  if (raw === undefined || raw === null) return undefined;
  const fail = (why: string): undefined => { log?.(`sessionFailure ignored (${why}): ${JSON.stringify(raw)}`); return undefined; };
  if (typeof raw !== 'object') return fail('not an object');
  const p = raw as Record<string, unknown>;
  const id = text(p.id);
  const title = text(p.title);
  const revision = typeof p.revision === 'number' && Number.isInteger(p.revision) && p.revision > 0 ? p.revision : undefined;
  const severity = p.severity === 'warning' || p.severity === 'error' ? p.severity : undefined;
  if (!id) return fail('id missing');
  if (revision === undefined) return fail(`revision not a positive integer: ${String(p.revision)}`);
  if (!severity) return fail(`severity invalid: ${String(p.severity)}`);
  if (!title) return fail('title missing');
  if (!Array.isArray(p.actions)) return fail('actions not an array');
  const category: FailureCategory = typeof p.category === 'string'
    ? (CATEGORIES.has(p.category as FailureCategory) ? p.category as FailureCategory : 'unknown')
    : 'unknown';
  const actions = [...new Set(p.actions.filter((a): a is FailureAction => ACTIONS.has(a as FailureAction)))];
  const details = text(p.details) ?? text(p.reason);
  return { id, revision, category, severity, title, ...(details ? { details } : {}), actions };
}

// The TurnError a turn-ending failure settles with: title (+ details) as the message, the category
// as the kind the alert card's copy line already prints, and the declared actions so the UI shows
// exactly the buttons the adapter offered
export function failureTurnError(f: SessionFailure): TurnError {
  return {
    message: f.details ? `${f.title}\n${f.details}` : f.title,
    kind: f.category,
    retryable: f.actions.includes('retry'),
    failureId: f.id,
    // Always present, even empty: an empty list means the adapter offers no remedy (quota, invalid
    // request), and the alert card must not fall back to its generic Retry / Reconnect pair
    actions: f.actions,
  };
}

File: rust/crates/acpira-host/src/acp/session_failure.rs
//! JetBrains AIR sessionFailure: structured warnings / errors at
//! `_meta.jetbrains.air.sessionFailure`; same id + higher revision updates in place

use serde_json::Value;

use acpira_shared::transcript::{FailureAction, FailureCategory, Severity, TurnError};

#[derive(Debug, Clone, PartialEq)]
pub struct SessionFailure {
  pub id: String,
  pub revision: f64,
  pub category: FailureCategory,
  pub severity: Severity,
  pub title: String,
  pub details: Option<String>,
  pub actions: Vec<FailureAction>,
}

fn text(v: Option<&Value>) -> Option<String> {
  v.and_then(Value::as_str).map(str::trim).filter(|s| !s.is_empty()).map(str::to_owned)
}

fn js_string(v: Option<&Value>) -> String {
  match v {
    None => "undefined".into(),
    Some(Value::String(s)) => s.clone(),
    Some(other) => other.to_string(),
  }
}

/// Strict read of one payload; None when absent or malformed (malformed is logged)
pub fn failure_of(meta: Option<&Value>, log: Option<&dyn Fn(&str)>) -> Option<SessionFailure> {
  let raw = meta?.get("jetbrains")?.get("air")?.get("sessionFailure")?;
  if raw.is_null() {
    return None;
  }
  let fail = |why: String| -> Option<SessionFailure> {
    if let Some(l) = log {
      l(&format!("sessionFailure ignored ({why}): {raw}"));
    }
    None
  };
  let Some(p) = raw.as_object() else { return fail("not an object".into()) };
  let id = text(p.get("id"));
  let title = text(p.get("title"));
  let revision = p.get("revision").and_then(Value::as_f64).filter(|r| r.fract() == 0.0 && *r > 0.0);
  let severity = match p.get("severity").and_then(Value::as_str) {
    Some("warning") => Some(Severity::Warning),
    Some("error") => Some(Severity::Error),
    _ => None,
  };
  let Some(id) = id else { return fail("id missing".into()) };
  let Some(revision) = revision else { return fail(format!("revision not a positive integer: {}", js_string(p.get("revision")))) };
  let Some(severity) = severity else { return fail(format!("severity invalid: {}", js_string(p.get("severity")))) };
  let Some(title) = title else { return fail("title missing".into()) };
  let Some(actions_raw) = p.get("actions").and_then(Value::as_array) else { return fail("actions not an array".into()) };
  let category =
    p.get("category").and_then(|c| serde_json::from_value::<FailureCategory>(c.clone()).ok()).unwrap_or(FailureCategory::Unknown);
  let mut actions: Vec<FailureAction> = vec![];
  for a in actions_raw {
    if let Ok(a) = serde_json::from_value::<FailureAction>(a.clone())
      && !actions.contains(&a)
    {
      actions.push(a);
    }
  }
  let details = text(p.get("details")).or_else(|| text(p.get("reason")));
  Some(SessionFailure { id, revision, category, severity, title, details, actions })
}

fn category_name(c: FailureCategory) -> String {
  serde_json::to_value(c).ok().and_then(|v| v.as_str().map(str::to_owned)).unwrap_or_default()
}

/// The TurnError a turn-ending failure settles with; actions always present, even empty
pub fn failure_turn_error(f: &SessionFailure) -> TurnError {
  TurnError {
    message: match &f.details {
      Some(d) => format!("{}\n{d}", f.title),
      None => f.title.clone(),
    },
    code: None,
    kind: Some(category_name(f.category)),
    retryable: Some(f.actions.contains(&FailureAction::Retry)),
    failure_id: Some(f.id.clone()),
    actions: Some(f.actions.clone()),
  }
}
